### PR TITLE
Fix a lot of CLOSE_WAIT connections.

### PR DESCRIPTION
**Pull request by sirkonst**
If `eof_received()` returns True (in this case) connection isn't really closing, which leads to the accumulation of a large number of open connections.

Step to reproduce:

1. Run server like:
```python
import asyncio
import logging

from aiosmtpd.handlers import Sink
from aiosmtpd.controller import Controller


async def amain(loop):
    cont = Controller(Sink(), hostname='::0', port=8025)
    cont.start()


if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG)
    loop = asyncio.get_event_loop()
    loop.create_task(amain(loop=loop))
    loop.run_forever()
```

2. Send several mails like:
```python
from smtplib import SMTP

client = SMTP( '127.0.0.1', 8025)

client.sendmail('from@test', ['to@test'], 'Test mail')
```

3. Check open connections:
```bash
$ lsof -n -P -p <PID>
```

### CODE REVIEW

**Comment by mosquito** [comment]
> Can you think of a way to add a unittest for this?

I guess, you have to mock linux tcp stack for this.

**Comment by sirkonst** [comment]
Test added :-)

**Comment by warsaw** [comment]
Does the new `test_connection_close()` test pass for you?  I get a consistent failure (after locally rebasing on master):

```
======================================================================
ERROR: test_connection_close (aiosmtpd.tests.test_smtp.TestSMTP)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/barry/projects/aiosmtpd/aiosmtpd/tests/test_smtp.py", line 446, in test_connection_close
    client.connect()
  File "/usr/lib/python3.5/smtplib.py", line 335, in connect
    self.sock = self._get_socket(host, port, self.timeout)
  File "/usr/lib/python3.5/smtplib.py", line 306, in _get_socket
    self.source_address)
  File "/usr/lib/python3.5/socket.py", line 712, in create_connection
    raise err
  File "/usr/lib/python3.5/socket.py", line 703, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
```

File: aiosmtpd/smtp.py [hunk 1]
```diff
@@ -149,6 +152,8 @@ def _handle_client(self):
                 continue
             yield from method(arg)
 
+            yield  # give a change to switch to another coroutine
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: Small coding style nits: I prefer to use non-hanging full sentences for comments (e.g. starts with a capital letter, ends in a period).  Also, no extraneous vertical whitespace.  Could you move the comment up to the preceding blank line and make it a full sentence?

File: aiosmtpd/smtp.py [hunk 2]
```diff
@@ -423,6 +428,8 @@ def smtp_DATA(self, arg):
                 data.append(line.decode('utf-8'))
             else:
                 data.append(line)
+
+            yield  # give a change to switch to another coroutine
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: Same comment here as above.

**Comment by warsaw** [comment]
Oh also, if you rebase on master, the CI checks should succeed (once the failing test if fixed).

**Comment by kozzztik** [comment]
Thats because no "close" called after handler coroutine finished. Saw that on production - in real usage many sockets are not closed properly. I have a fix for this in #4, but not sure I have a test for it. May be I will split my PR to small ones as there are some fixed bugs and PR it too huge to be merged at once.

**Comment by mosquito** [comment]
We are already tested it. That's working fine. So, we are have to fork this repo for using unapproved changes in production. I hope @warsaw will merge it as soon as possible and we are starting a progress of moving to this project. Otherwise we are have to support the fork.

**Comment by warsaw** [comment]
@mosquito For sure I am very willing to merge this and other branches, if we can get them passing CI, clean up a few coding nits, and get at least 100% coverage for new code (see `tox -e py35-diffcov` for example).  I would like to avoid forks!

I think if we can get these last few PRs in, we can release a 1.0 final.

File: aiosmtpd/smtp.py [hunk 3]
```diff
@@ -132,8 +132,11 @@ def _client_connected_cb(self, reader, writer):
         self._writer = writer
 
     def eof_received(self):
+        super().eof_received()
```
Comment (warsaw) [inline aiosmtpd/smtp.py]: This looks weird.  Why do you need this overridden method at all if you're just going to upcall to super?

File: aiosmtpd/controller.py [hunk 4]
```diff
@@ -42,12 +42,12 @@ def _run(self, ready_event):
         sock = self.make_socket()
         sock.bind((self.hostname, self.port))
         asyncio.set_event_loop(self.loop)
-        server = self.loop.run_until_complete(
+        self.server = self.loop.run_until_complete(
```
Comment (warsaw) [inline aiosmtpd/controller.py]: Why keep the server instance on `self`?

File: aiosmtpd/tests/test_smtp.py [hunk 5]
```diff
@@ -52,10 +53,10 @@ def process_message(self, peer, mailfrom, rcpttos, data, **kws):
 
 class TestSMTP(unittest.TestCase):
     def setUp(self):
-        controller = UTF8Controller(Sink)
-        controller.start()
-        self.addCleanup(controller.stop)
-        self.address = (controller.hostname, controller.port)
+        self.controller = UTF8Controller(Sink)
```
Comment (warsaw) [inline aiosmtpd/tests/test_smtp.py]: Why keep the controller instance on `self`?

**Comment by warsaw** [comment]
I have a number of questions about the code, and some minor fix requests about style.  Please address these and also make sure the tests pass.  I see lots of failures with this branch.  I tried to merge it with master via the GH web ui, so hopefully I didn't mess that up.  Please review.

**Comment by warsaw** [comment]
Is this branch still relevant?  There have been a bunch of changes and fixes in the meantime, and there are now a lot of conflicts.  I'll assume this branch isn't needed any more, but if you think it still is, please reopen and resolve, or create a new PR.  Thanks!

**Pull request closed without merging**